### ISSUE TITLE
False positive on `VarCouldBeVal` in generic classes

### ISSUE THREAD

**Issue by rli** [detekt/detekt#4730]
## Observed Behavior
`Expecting empty but was: [CodeSmell(issue=Issue(id='VarCouldBeVal', severity=Maintainability, debt=5min), entity=Entity(name=a, signature=Test.kt$A$private var a = 1, location=Location(source=2:5, text=17:34, file=Test.kt, filePath=FilePath(absolutePath=Test.kt, basePath=null, relativePath=null)), ktElement=PROPERTY), message=Variable 'a' could be val., metrics=[], references=[], severity=WARNING, id='VarCouldBeVal')]`

## Steps to Reproduce
```kt
        @Test
        fun `does not report variables that are re-assigned in generic class with receiver`() {
            val code = """
                class A<T> {
                    private var a = 1
                    
                    fun foo(): A<T> = apply {
                        a = 2
                    }
                }
            """.trimIndent()
            assertThat(subject.compileAndLintWithContext(env, code)).isEmpty()
        }
```

## Context

Interestingly, declaring the return type as generic and the fact that it's in a receiver block are important to getting the lint error to report.
```kt
class A<T> {
    private var a = 1

    // fun foo() = apply { a = 2 }         // no error
    // fun foo2(): A = apply { a = 3 }     // no error

    fun foo3(): A<T> = apply { a = 4 }  // error

    // no error
    /*
    fun foo4(): A<T> {
        a = 5
        return this
    }
    */
}
```

## Your Environment

* Version of detekt used: 1.20.0
* Link to your project (if it's a public repository): https://github.com/aws/aws-toolkit-jetbrains/blob/7853ede42bb44a585b786b0ad49c81d27d24c9cc/jetbrains-core/src/software/aws/toolkits/jetbrains/ui/ResourceSelector.kt#L232-L267
